File: Intersection of Two Linked Lists LC-160.py
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if(!headA || !headB)
            return NULL;
        ListNode* temp1=headA;
        ListNode* temp2=headB;
        
        /***Two Pointers***/
        
       // while(temp1!=temp2){
       //     if(temp1==NULL)
       //         temp1=headB;
       //     else 
       //         temp1=temp1->next;
       //     if(temp2==NULL)
       //         temp2=headA;
       //     else
       //         temp2=temp2->next;
       // }

       while(temp1!=temp2){
           temp1=temp1==NULL?headB:temp1->next;
           temp2=temp2==NULL?headA:temp2->next;
        }
        return temp1;
    }
};
//Simple One
/*
  struct ListNode* tempA=headA;
        struct ListNode* tempB=headB;
        while(tempA != NULL) {
            while(tempB != NULL) {
                if(tempA == tempB) {
                    return tempA;
                }
                tempB = tempB->next;
            }
            tempB = headB;
            tempA = tempA->next;
        }
        return NULL;
*/


//Simple looping
/*
ListNode* temp1=headA;
        ListNode* temp2=headB;
        if(headA==headB)
            return headA;
        while(temp1 && temp2){
            if(temp1==temp2->next)
                return temp1;
            else if(temp1->next==temp2)
                return temp2;
            else if(temp1->next==temp2->next)
                return temp1->next;
            temp1=temp1->next;
            temp2=temp2->next;
        }
        return NULL;
*/

//Hashing
/*
  ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        
      ListNode* temp1=headA;
        unordered_set<ListNode*> set;
        while(temp1){
            set.insert(temp1);
            temp1=temp1->next;
        }
        ListNode* temp2=headB;
        while(temp2){
            if(set.find(temp2)!=set.end())
                return temp2;
            temp2=temp2->next;
        }
        return NULL;
*/

File: Intersection of Two Linked Lists LC160.cpp
// Same thing as the two next but code modified
// tc=o(n)+o(n)
// sc=o(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if(!headA || !headB) return NULL;

        ListNode *d1=headA,*d2=headB;
        while(d1!=d2){
            d1=d1?d1->next:headB;
            d2=d2?d2->next:headA;
        }
        return d1;
    }
};

// Same thing as the onenext but code modified, just swapped
// tc=o(n)+o(n)
// sc=o(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *d1=headA, *d2=headB;
        while(d1 && d2){
            d1=d1->next;
            d2=d2->next;
        }

        if(d1){
            swap(d1,d2);
            swap(headA,headB);
        }

        d1=headB;
        while(d2){
            d1=d1->next;
            d2=d2->next;
        }
        d2=headA;

        while(d1 && d2){
            if(d1==d2) return d1;
            d1=d1->next;
            d2=d2->next;
        }
        return NULL;
    }
};

// tc=o(n)+o(n)
// sc=o(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *d1=headA, *d2=headB;
        while(d1 && d2){
            d1=d1->next;
            d2=d2->next;
        }
        if(d1){
            d2=headA;
            while(d1){
                d1=d1->next;
                d2=d2->next;    
            }
            d1=headB;
        }else{
            d1=headB;
            while(d2){
                d1=d1->next;
                d2=d2->next;
            }
            d2=headA;
        }
        while(d1 && d2){
            if(d1==d2) return d1;
            d1=d1->next;
            d2=d2->next;
        }
        return NULL;
    }
};


// tc=o(n)+o(n)
// sc=o(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int c1=0,c2=0;
        ListNode *d1=headA, *d2=headB;
        while(d1 && d2){
            d1=d1->next;
            d2=d2->next;
            c1++;
            c2++;
        }
        while(d1){
            d1=d1->next;
            c1++;           
        }
        while(d2){
            d2=d2->next;
            c2++;           
        }
        
        d1=headA;
        d2=headB;
        if(c1<c2){
            swap(c1,c2);
            swap(d1,d2);
        }
        int diff=c1-c2;
        while(diff--){
            d1=d1->next;
        }
        while(d1 && d2){
            if(d1==d2) return d1;
            d1=d1->next;
            d2=d2->next;
        }
        return NULL;
    }
};


// hashset
// tc=o(m)
// sc=o(n)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        unordered_set<ListNode*> set;
        while(headA){
            set.insert(headA);
            headA=headA->next;
        }
        while(headB){
            if(set.find(headB)!=set.end()){
                return headB;
            }
            headB=headB->next;
        }
        return NULL;
    }
};